### PR TITLE
Add Index to ENR when used on temp tables, prevent temp table index recreation issue 

### PR DESCRIPTION
**Pull request by time-change**
### Description

With this change, we add newly created indexes to ENR when used on temp tables, to match Table Variable behavior in these cases.

In addition, we add an extra scan in findDependentObjects to properly clean up index objects left in catalog when used with temp tables. Previously, entries would be left in pg_catalog, which would prevent recreating the same table with the same index name. This change also prevents rare cases of dangling catalog entries in cases where we aren't able to clean up properly, such as in a crash.
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/catalog/dependency.c [hunk 1]
```diff
@@ -950,9 +952,85 @@ findDependentObjects(const ObjectAddress *object,
 		dependentObjects[numDependentObjects].subflags = subflags;
 		numDependentObjects++;
 	}
-
+	is_enr = scan->enr;
 	systable_endscan(scan);
 
+	/* Indices for temp tables will leave metadata in pg_catalog, so we should clean them up here. */
+	if (is_enr)
```
Comment (lejaokri) [inline src/backend/catalog/dependency.c]: I am wondering why is this needed if code is supposed to add temp table index to ENR? If index is in ENR then we wouldnt need this, no?

File: src/backend/catalog/index.c [hunk 2]
```diff
@@ -750,8 +750,10 @@ index_create(Relation heapRelation,
  		 * Index of Table Variables should also be in ENR and the index
 		 * can only be created as part of DECLARE @[name] TABLE;
 		 * when specifying PRIMARY KEYs and/or CONSTRAINTs.
+		 *
+		 * Indexes of temp tables should be in enr as well, to prevent cache issues.
 		 */
-		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && indexRelationName[0] == '@');
+		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && (indexRelationName[0] == '@' || indexRelationName[0] == '#'));
```
Comment (lejaokri) [inline src/backend/catalog/index.c]: In `heap_create_with_catalog()` you have a check there so not all temp tables are added in ENR but here you are doing all "index" of temp tables should be in enr. What happens in the case where the temp table is NOT in enr but index is in enr?

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (time-change) [reply src/backend/catalog/dependency.c]: This is needed because the below code does add the pg_index entries to ENR, but doesn't prevent them from being created in catalogs. It's those entries in pg_class, pg_depend etc that will not be cleaned up in findDependentObjects.

File: src/backend/catalog/index.c (hunk 2 above)
Comment (time-change) [reply src/backend/catalog/index.c]: In that case, this line might be unnecessary.

Comment (surendravishnoi) [reply src/backend/catalog/index.c]: we need to follow all in ENR or nothing. We should put indexes in ENR if corresponding tables are in ENR. We should have same check as that for tables.

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (surendravishnoi) [reply src/backend/catalog/dependency.c]: Can we add indexes to ENR only? If we are keeping relations in ENR only, why not indexes?

File: src/backend/catalog/index.c (hunk 2 above)
Comment (time-change) [reply src/backend/catalog/index.c]: agreed, reworked to handle indexes in ENR using same check from heap_create_with_catalog.

File: src/backend/catalog/index.c [hunk 3]
```diff
@@ -750,8 +751,16 @@ index_create(Relation heapRelation,
...
+		 *
+		 * Indexes of temp tables should be in enr as well, to prevent cache issues.
+		 * Index names can actually contain '#' naturally, but because of construct_unique_index_name
+		 * we have no way of checking that without the original index name provided to parser.
 		 */
-		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && indexRelationName[0] == '@');
+		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && (indexRelationName[0] == '@' || strchr(indexRelationName, '#') != NULL));
+
+		/* One last check, if the table has any dependencies */
+		if(CheckTempTableHasDependencies(heapRelation->rd_att))
```
Comment (lejaokri) [inline src/backend/catalog/index.c]: I can think of two issues:

1. This isnt completely consistent with the check in heap_create_with_catalog(). This check here doesnt ignore Table Variables but the one in heap_create_with_catalog() ignores Table Variables. The reason is that we want everything associated to TVs in ENR.

2. Consider the following test
-> create table where CheckTempTableHasDependencies returns false this means table is in ENR
-> Alter table to include a column where CheckTempTableHasDependencies would return True
-> Create an index, now CheckTempTableHasDependencies would return True so you will end up with table in ENR, and index not in ENR.

Comment (time-change) [reply src/backend/catalog/index.c]: for case 2), i don't know if there's a simple solution for that situation. i think the problem then is more with the alter table, and not the index creation, right? in theory, we would have to do something like move ENR metadata out in that scenario.

I think that fix would be out of scope for this issue.

Comment (lejaokri) [reply src/backend/catalog/index.c]: Perhaps `get_ENR_withoid()` can help

File: src/backend/catalog/index.c [hunk 4]
```diff
@@ -747,11 +748,14 @@ index_create(Relation heapRelation,
...
-		 * when specifying PRIMARY KEYs and/or CONSTRAINTs.
+		 * Index of temp objects in ENR should also be in ENR.
+		 * For table variables, index can only be created as
+		 * part of DECLARE @[name] TABLE; when specifying
+		 * PRIMARY KEYs and/or CONSTRAINTs.
 		 */
-		is_enr = (indexRelationName && strlen(indexRelationName) > 0 && indexRelationName[0] == '@');
+		is_enr = (indexRelationName && strlen(indexRelationName) > 0 &&
+			(indexRelationName[0] == '@' || (strchr(indexRelationName, '#') != NULL
+			&& get_ENR_withoid(currentQueryEnv, heapRelationId, ENR_TSQL_TEMP))));
```
Comment (lejaokri) [inline src/backend/catalog/index.c]: nitpick for performance purposes we do not need to call `get_ENR_withoid()` if it is a table variable.

Comment (time-change) [reply src/backend/catalog/index.c]: I don't think it's being called if it's a TV, there's parens around strchr+get_ENR_withoid

Comment (lejaokri) [reply src/backend/catalog/index.c]: okay got it thanks thats a bit painful to read :)

Comment (lejaokri) [reply src/backend/catalog/index.c]: Are we also allowed to create index for temp tables where the index name does not start with '#'?

Comment (time-change) [reply src/backend/catalog/index.c]: The strchr call is unnecessary, yeah

**Review by lejaokri — APPROVED** [review]

**Review by priyansx — APPROVED** [review]
LGTM

File: src/backend/catalog/dependency.c [hunk 5]
```diff
@@ -76,6 +76,7 @@
 #include "commands/typecmds.h"
 #include "funcapi.h"
 #include "nodes/nodeFuncs.h"
+#include "parser/parser.h"
```
Comment (surendravishnoi) [inline src/backend/catalog/dependency.c:79]: why this change?
